Adding a message type for Log messages that will be sent out over the…

nburek opened the pull request:
Adding a message type for Log messages that will be sent out over the rosout topic.

Issue for this request: https://github.com/ros2/rcl_interfaces/issues/50

nburek committed:
Adding a message type for Log messages that will be sent out over the rosout topic.

cr https://code.amazon.com/reviews/CR-3890422

tfoote commented:
For reference this is a migration of the rosgraph_msgs/Log in ROS1: http://docs.ros.org/melodic/api/rosgraph_msgs/html/msg/Log.html
```
##
## Severity level constants
##
byte DEBUG=1 #debug level
byte INFO=2  #general level
byte WARN=4  #warning level
byte ERROR=8 #error level
byte FATAL=16 #fatal/critical level
##
## Fields
##
Header header
byte level
string name # name of the node
string msg # message 
string file # file the message came from
string function # function the message came from
uint32 line # line the message came from
string[] topics # topic names that the node publishes
```


The primary changes are changing the constant values and dropping the topics. 
I agree that dropping the topics makes sense. 

The enum values have changed from a bitmask like approach to a sequential approach with padding. This is matching the values in rcutils: https://github.com/ros2/rcutils/blob/master/include/rcutils/logging.h#L156-L165
They were set in https://github.com/ros2/rcutils/pull/57 to follow python's conventions: https://github.com/ros2/rcutils/pull/57#pullrequestreview-69952305

Since we're using an enum and I don't believe that we're planning to support custom logging levels I'd suggest that we restore the old enum values for consistency adding 0 for UNSET.

dirk-thomas commented:
> name of the node

This comment is only partially true. The name might be a node name but it can also be any other user defined string.

> Since we're using an enum and I don't believe that we're planning to support custom logging levels I'd suggest that we restore the old enum values.

The logging system already does support arbitrary numbers in between. So the constants should match the implementation exactly. The type should be changed from `byte` to `uint8` though since the data is actually numeric with its implicit order and not an opaque octet.

> for consistency adding 0 for UNSET.

Since there is no semantic associated to `0` atm I think there shouldn't be a constant defined in the message for it.

nburek committed:
Changing the type for level from byte to uint8 and updating the comment for the name field.

nburek commented:
> > name of the node
> 
> This comment is only partially true. The name might be a node name but it can also be any other user defined string.

I'll update the comment to mention that the field is meant to represent the name of the logger used to log this message. The logger name won't exactly match the node name anyways


> > Since we're using an enum and I don't believe that we're planning to support custom logging levels I'd suggest that we restore the old enum values.
> 
> The logging system already does support arbitrary numbers in between. So the constants should match the implementation exactly. The type should be changed from `byte` to `uint8` though since the data is actually numeric with its implicit order and not an opaque octet.
> 
> >

I'll change the type from `byte` to `uint8`.

mlautman commented:
Don't forget to add msg/Log.msg to the rosidl_generate_interfaces call

mlautman commented:
also, you need to replace `Header` with `common_interfaces/Header`

mlautman commented:
Header -> std_msgs/Header

nburek committed:
Adding the new Log msg to the CMake file and removing the Header from the Log.

nburek commented:
Thanks for the feedback @mlautman. I've added the Log.msg to the rosidl_generate_interface call in CMake. I've also removed the header from the Log message for now since it would add a dependency on std_msgs from common_interfaces. In the conversation on issue https://github.com/ros2/rcl_interfaces/issues/50, there was a desire to not have any new dependencies added in front of rcl.

mlautman commented:
Seems strange to not be able to depend on a core library like std_msgs. The header field is really useful

dirk-thomas commented:
> Seems strange to not be able to depend on a core library like std_msgs.

That statement is taken out of context. It is undesired to add a dependency to the `common_interfaces` package into the `rcl` packages. Since the log message should continue to use a header field the `std_msgs` package (or maybe just that specific message) could be moved into the `rcl_interfaces` repo...

nburek commented:
It may make sense for the Header message definition to be moved into rcl_interfaces. If that's the case then the Log message could add it back in as a field. However, should that be done for this pull request or should we move forward with the Log message as defined in this request and then create a new issue for the follow up work of moving either `std_msgs` or `Header` so they are part of the core dependencies?

dirk-thomas commented:
The `Header` also contains a frame id which isn't needed / used here. Therefore I would suggest to add a timestamp field instead.

nburek commented:
> The `Header` also contains a frame id which isn't needed / used here. Therefore I would suggest to add a timestamp field instead.

That may beg the question of if frame id belongs inside of `Header` or not. Judging by the comment in the .msg file, the frame id pertains to transformation data. Does it make sense to remove the frame id to make `Header` apply more generally or should header potentially be moved/renamed so that it's more apparent the Header is meant to be used with Transform messages.

wjwwood commented:
> That may beg the question of if frame id belongs inside of Header or not. Judging by the comment in the .msg file, the frame id pertains to transformation data. Does it make sense to remove the frame id to make Header apply more generally or should header potentially be moved/renamed so that it's more apparent the Header is meant to be used with Transform messages.

I would say no... The header is used mostly in the sensor messages where data both has a creation time and a physical origin location, in which case the frame id makes sense.

I think it's just convenient for people to grab the header when they needed a timestamp and/or a sequence number (sequence number is no longer in ROS 2's header message). Personally, I think this is a case of "Header abuse".

tfoote commented:
The std_msgs/Header has significant meaning for physically connected data which is almost all data at the robotic developer level or robot application level.  What we're doing here is much more internal to the middleware. We could define a new datatype but we've already deprecated the sequence number in the header and if you remove the frame_id it's just a timestamp. So just using a raw timestamp is actually really all that there is to a std_msgs/Header if we don't want the frame_id.

There's probably a slightly better name for std_msgs/Header but since it's used throughout the ROS ecosystem in it's current name, changing it is much more of a hassle than it's worth.

tfoote commented on rcl_interfaces/msg/Log.msg:
```diff
@@ -0,0 +1,18 @@
+
+##
+## Fields
+##
```
```suggestion
##
builtin_interfaces/Time stamp
```

nburek commented:
Alright, good to know. I'll go ahead and add timestamp as a Time type field

tfoote committed:
Adding a timestamp to the message in lieu of Header, which has fields that aren't applicable to Log messages

Co-Authored-By: nburek <NickBurek@gmail.com>

tfoote committed:
Add newly added message dependency builtin_interfaces

Added for the Time datatype.

mlautman commented on rcl_interfaces/CMakeLists.txt:
```diff
@@ -18,6 +18,7 @@ find_package(rosidl_default_generators REQUIRED)
 rosidl_generate_interfaces(${PROJECT_NAME}
```
```suggestion
find_package(builtin_interfaces REQUIRED)

rosidl_generate_interfaces(${PROJECT_NAME}
```

mlautman commented:
Add `find_package(builtin_interfaces REQUIRED)` after `find_package(ament_cmake REQUIRED)` and `  DEPENDENCIES builtin_interfaces` at the end of `rosidl_generate_interfaces`

nburek commented:
> Add `find_package(builtin_interfaces REQUIRED)` after `find_package(ament_cmake REQUIRED)` and ` DEPENDENCIES builtin_interfaces` at the end of `rosidl_generate_interfaces`

Thanks Mike. I'm actually in the middle of doing that now and verifying it builds locally. Will update pull request in a minute.

nburek committed:
Adding builtin_interfaces as a dependency in rcl_interfaces' cmake file.

tfoote commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5668)](http://ci.ros2.org/job/ci_linux/5668/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2276)](http://ci.ros2.org/job/ci_linux-aarch64/2276/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4686)](http://ci.ros2.org/job/ci_osx/4686/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5561)](http://ci.ros2.org/job/ci_windows/5561/)

tfoote reviewed (approved):
This lgtm and it's passing CI.
